File: indexer/indexer/observability/otel.py
from __future__ import annotations

import json
import logging
import os
import secrets
import socket
import time
from dataclasses import dataclass, field
from typing import Any, Mapping

from opentelemetry import baggage, trace
from opentelemetry.baggage.propagation import W3CBaggagePropagator
from opentelemetry.context import Context
from opentelemetry.propagators.composite import CompositePropagator
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.trace import (
    NonRecordingSpan,
    SpanContext,
    SpanKind,
    Status,
    StatusCode,
    TraceFlags,
)
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

logger = logging.getLogger(__name__)

SCOPE_NAME = "ton-indexer.observability"
SCOPE_VERSION = "1.0.0"
OTEL_DATA_FIELD = "otel_data"

ACTION_CLASSIFIER_SERVICE_NAME = "action-classifier"
ACTION_CLASSIFIER_SPAN_NAME = "ton.action_classifier.process"
ACTION_CLASSIFIER_SERVICE_STAGE = "action_classifier"

PROPAGATION_BAGGAGE_KEYS = (
    "ton.trace.external_message_hash_norm",
    "ton.trace.external_message_hash",
    "ton.trace.finality",
    "ton.processing.pass_id",
    "ton.trace.root_tx_hash",
)

_TRUE_VALUES = {"1", "true", "yes", "on"}
_PROPAGATOR = CompositePropagator(
    [TraceContextTextMapPropagator(), W3CBaggagePropagator()]
)
_TRACER = None
_SERVICE_INSTANCE_ID = socket.gethostname().strip()


def _flag(name: str, default: bool = False) -> bool:
    raw = os.getenv(name)
    if raw is None:
        return default
    return raw.strip().lower() in _TRUE_VALUES


def tracing_enabled() -> bool:
    return (
        _flag("TON_OTEL_ENABLED")
        and not _flag("OTEL_SDK_DISABLED")
        and os.getenv("OTEL_TRACES_EXPORTER", "otlp").strip().lower() != "none"
    )


def measure_logs_enabled() -> bool:
    return False


def _build_exporter():
    protocol = (
        os.getenv("OTEL_EXPORTER_OTLP_TRACES_PROTOCOL")
        or os.getenv("OTEL_EXPORTER_OTLP_PROTOCOL")
        or "http/protobuf"
    ).strip().lower()

    if protocol in {"http", "http/protobuf", ""}:
        from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
        return OTLPSpanExporter()

    if protocol == "grpc":
        from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import (
            OTLPSpanExporter as OTLPGrpcSpanExporter,
        )
        return OTLPGrpcSpanExporter()

    raise ValueError(f"Unsupported OTLP protocol: {protocol}")

def init_tracing() -> None:
    global _TRACER
    if _TRACER is not None or not tracing_enabled():
        return

    try:
        resource_attributes = {
            "service.namespace": "ton",
            "service.name": ACTION_CLASSIFIER_SERVICE_NAME,
        }
        if _SERVICE_INSTANCE_ID:
            resource_attributes["service.instance.id"] = _SERVICE_INSTANCE_ID

        provider = TracerProvider(
            resource=Resource.create(resource_attributes)
        )
        provider.add_span_processor(BatchSpanProcessor(_build_exporter()))
        _TRACER = provider.get_tracer(SCOPE_NAME, SCOPE_VERSION)
    except Exception:
        logger.exception(
            "Failed to initialize tracing for %s",
            ACTION_CLASSIFIER_SERVICE_NAME,
        )


def new_pass_id() -> str:
    return secrets.token_hex(16)


def _normalized_pass_id(value: Any) -> str:
    if not isinstance(value, str):
        return ""

    pass_id = value.strip().lower()
    if len(pass_id) != 32:
        return ""

    try:
        trace_id = int(pass_id, 16)
    except ValueError:
        return ""

    if trace_id == 0:
        return ""

    return pass_id


def _root_context_from_pass_id(pass_id: str, base_context: Any) -> Any:
    normalized_pass_id = _normalized_pass_id(pass_id)
    if not normalized_pass_id:
        return base_context

    trace_id = int(normalized_pass_id, 16)
    span_id = secrets.randbits(64) or 1
    span_context = SpanContext(
        trace_id=trace_id,
        span_id=span_id,
        is_remote=True,
        trace_flags=TraceFlags(0x01),
    )
    return trace.set_span_in_context(NonRecordingSpan(span_context), base_context)


def _extract_stage_context(
    trace_hash: Mapping[str, Any],
) -> tuple[Any, dict[str, str], bool]:
    raw = trace_hash.get(OTEL_DATA_FIELD)
    if isinstance(raw, bytes):
        raw = raw.decode("utf-8")

    payload = {}
    if raw:
        try:
            payload = json.loads(str(raw))
        except (TypeError, json.JSONDecodeError):
            pass

    carrier = {
        key: str(value)
        for key in ("traceparent", "tracestate", "baggage")
        if (value := payload.get(key))
    }

    parent_context = _PROPAGATOR.extract(carrier=carrier) if carrier else Context()
    baggage_values = {
        key: str(value)
        for key, value in baggage.get_all(context=parent_context).items()
    }
    context_missing = not trace.get_current_span(parent_context).get_span_context().is_valid

    return parent_context, baggage_values, context_missing


def _redis_otel_fields(context: Any, attributes: Mapping[str, Any]) -> dict[str, str]:
    for key in PROPAGATION_BAGGAGE_KEYS:
        value = attributes.get(key)
        if value not in (None, ""):
            context = baggage.set_baggage(key, str(value), context=context)

    carrier: dict[str, str] = {}
    _PROPAGATOR.inject(carrier=carrier, context=context)

    return {
        OTEL_DATA_FIELD: json.dumps(
            {
                "traceparent": carrier.get("traceparent", ""),
                "tracestate": carrier.get("tracestate", ""),
                "baggage": carrier.get("baggage", ""),
            },
            separators=(",", ":"),
        )
    }


@dataclass(slots=True)
class StageSpan:
    attributes: dict[str, Any] = field(default_factory=dict)
    context: Any = None
    span: Any = None
    error_type: str | None = None
    error_message: str | None = None
    ended: bool = False

    def add_attr(self, key: str, value: Any | None) -> None:
        if value not in (None, ""):
            self.attributes[key] = value

    def mark_error(self, error_type: str, message: str | None = None) -> None:
        self.error_type = error_type
        self.error_message = message
        self.attributes["ton.error.type"] = error_type

    def propagation_fields(self) -> dict[str, str]:
        if self.span is None:
            return {}
        return _redis_otel_fields(self.context, self.attributes)

    def emit(self) -> None:
        if self.span is None or self.ended:
            return

        self.span.set_attributes(self.attributes)

        if self.error_type:
            self.span.set_status(
                Status(StatusCode.ERROR, self.error_message or self.error_type)
            )

        self.span.end(end_time=time.time_ns())
        self.ended = True


def new_stage(*, start_time_ns: int, trace_hash: Mapping[str, Any]) -> StageSpan:
    parent_context, baggage_values, context_missing = _extract_stage_context(trace_hash)
    parent_span_context = trace.get_current_span(parent_context).get_span_context()

    attributes: dict[str, Any] = {
        "ton.processing.pipeline": "trace_to_actions_to_stream",
        "ton.processing.service_stage": ACTION_CLASSIFIER_SERVICE_STAGE,
        "transaction.type": "processing",
    }
    if _SERVICE_INSTANCE_ID:
        attributes["service.instance.id"] = _SERVICE_INSTANCE_ID

    for key in PROPAGATION_BAGGAGE_KEYS:
        if value := baggage_values.get(key):
            attributes[key] = value

    pass_id = _normalized_pass_id(attributes.get("ton.processing.pass_id"))
    if parent_span_context.is_valid:
        pass_id = format(parent_span_context.trace_id, "032x")
    if not pass_id:
        pass_id = new_pass_id()
    attributes["ton.processing.pass_id"] = pass_id

    if context_missing:
        attributes["ton.processing.context_missing"] = True
        parent_context = _root_context_from_pass_id(pass_id, parent_context)

    span = None
    context = parent_context

    if _TRACER is None and tracing_enabled():
        init_tracing()

    if _TRACER is not None:
        span = _TRACER.start_span(
            name=ACTION_CLASSIFIER_SPAN_NAME,
            context=parent_context,
            kind=SpanKind.CONSUMER,
            start_time=start_time_ns,
        )
        context = trace.set_span_in_context(span, parent_context)

    return StageSpan(attributes=attributes, context=context, span=span)

File: ton-streaming-go/observability/otel.go
package observability

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"log"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/baggage"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

const (
	scopeName             = "ton-indexer.observability"
	scopeVersion          = "1.0.0"
	otelDataField         = "otel_data"
	streamingServiceName  = "ton-streaming-api"
	streamingSpanName     = "ton.streaming_api.process"
	streamingServiceStage = "streaming_api"
)

var (
	propagationBaggageKeys = []string{
		"ton.trace.external_message_hash",
		"ton.trace.finality",
		"ton.processing.pass_id",
		"ton.trace.root_tx_hash",
	}

	redisPropagator = propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	)

	tracerOnce    sync.Once
	processTracer trace.Tracer
)

type otelData struct {
	Traceparent string `json:"traceparent"`
	Tracestate  string `json:"tracestate"`
	Baggage     string `json:"baggage"`
}

type StageSpan struct {
	Attributes   map[string]any
	ErrorType    string
	ErrorMessage string

	context context.Context
	span    trace.Span
	ended   bool
}

func NowUnixNano() int64 {
	return time.Now().UnixNano()
}

func TracingEnabled() bool {
	if !envFlag("TON_OTEL_ENABLED", false) {
		return false
	}
	if envFlag("OTEL_SDK_DISABLED", false) {
		return false
	}
	return strings.ToLower(strings.TrimSpace(os.Getenv("OTEL_TRACES_EXPORTER"))) != "none"
}

func MeasureLogsEnabled() bool {
	return envFlag("MEASURE_LOGS_ENABLED", false)
}

func NewStage(startTimeUnix int64, traceHash map[string]string) *StageSpan {
	parentContext, baggageValues, contextMissing := extractStageContext(traceHash)

	attributes := map[string]any{
		"ton.processing.pipeline":      "trace_to_actions_to_stream",
		"ton.processing.service_stage": streamingServiceStage,
	}

	for _, key := range propagationBaggageKeys {
		if value := baggageValues[key]; value != "" {
			attributes[key] = value
		}
	}

	if _, ok := attributes["ton.processing.pass_id"]; !ok {
		attributes["ton.processing.pass_id"] = randomHex(16)
	}

	if contextMissing {
		attributes["ton.processing.context_missing"] = true
	}

	stage := &StageSpan{
		Attributes: attributes,
		context:    parentContext,
	}

	tracer := currentTracer()
	if tracer == nil {
		return stage
	}

	spanContext, span := tracer.Start(
		parentContext,
		streamingSpanName,
		trace.WithSpanKind(trace.SpanKindInternal),
		trace.WithTimestamp(time.Unix(0, startTimeUnix)),
	)

	stage.context = spanContext
	stage.span = span
	return stage
}

func (s *StageSpan) AddAttr(key string, value any) {
	if s == nil || value == nil {
		return
	}
	if str, ok := value.(string); ok && str == "" {
		return
	}
	s.Attributes[key] = value
}

func (s *StageSpan) MarkError(errorType, message string) {
	if s == nil {
		return
	}
	s.ErrorType = errorType
	s.ErrorMessage = message
	s.Attributes["ton.error.type"] = errorType
}

func (s *StageSpan) PreparePropagationFields() map[string]string {
	if s == nil || s.span == nil {
		return nil
	}
	return redisOtelFields(s.context, s.Attributes)
}

func (s *StageSpan) Emit() {
	if s == nil || s.span == nil || s.ended {
		return
	}

	s.span.SetAttributes(attributesToKeyValues(s.Attributes)...)

	if s.ErrorType != "" {
		message := s.ErrorMessage
		if message == "" {
			message = s.ErrorType
		}
		s.span.SetStatus(codes.Error, message)
	}

	s.span.End()
	s.ended = true
}

func extractStageContext(traceHash map[string]string) (context.Context, map[string]string, bool) {
	raw := strings.TrimSpace(traceHash[otelDataField])

	var payload otelData
	if raw != "" {
		_ = json.Unmarshal([]byte(raw), &payload)
	}

	carrier := propagation.MapCarrier{}
	if payload.Traceparent != "" {
		carrier.Set("traceparent", payload.Traceparent)
	}
	if payload.Tracestate != "" {
		carrier.Set("tracestate", payload.Tracestate)
	}
	if payload.Baggage != "" {
		carrier.Set("baggage", payload.Baggage)
	}

	parentContext := redisPropagator.Extract(context.Background(), carrier)

	baggageValues := map[string]string{}
	for _, member := range baggage.FromContext(parentContext).Members() {
		baggageValues[member.Key()] = member.Value()
	}

	contextMissing := !trace.SpanContextFromContext(parentContext).IsValid()
	return parentContext, baggageValues, contextMissing
}

func redisOtelFields(ctx context.Context, attributes map[string]any) map[string]string {
	if ctx == nil {
		ctx = context.Background()
	}

	merged := map[string]baggage.Member{}
	for _, member := range baggage.FromContext(ctx).Members() {
		merged[member.Key()] = member
	}

	for _, key := range propagationBaggageKeys {
		value := strings.TrimSpace(asString(attributes[key]))
		if value == "" {
			continue
		}

		member, err := baggage.NewMember(key, value)
		if err != nil {
			continue
		}
		merged[key] = member
	}

	members := make([]baggage.Member, 0, len(merged))
	for _, member := range merged {
		members = append(members, member)
	}

	if len(members) > 0 {
		bg, err := baggage.New(members...)
		if err == nil {
			ctx = baggage.ContextWithBaggage(ctx, bg)
		}
	}

	carrier := propagation.MapCarrier{}
	redisPropagator.Inject(ctx, carrier)

	payload, err := json.Marshal(otelData{
		Traceparent: carrier.Get("traceparent"),
		Tracestate:  carrier.Get("tracestate"),
		Baggage:     carrier.Get("baggage"),
	})
	if err != nil {
		return nil
	}

	return map[string]string{
		otelDataField: string(payload),
	}
}

func currentTracer() trace.Tracer {
	if !TracingEnabled() {
		return nil
	}

	tracerOnce.Do(func() {
		processTracer = buildTracer()
	})

	return processTracer
}

func buildTracer() trace.Tracer {
	exporter := buildExporter()
	if exporter == nil {
		return nil
	}

	provider := sdktrace.NewTracerProvider(
		sdktrace.WithResource(resource.NewSchemaless(
			attribute.String("service.namespace", "ton"),
			attribute.String("service.name", streamingServiceName),
		)),
		sdktrace.WithSpanProcessor(sdktrace.NewBatchSpanProcessor(exporter)),
	)

	return provider.Tracer(
		scopeName,
		trace.WithInstrumentationVersion(scopeVersion),
	)
}

func buildExporter() sdktrace.SpanExporter {
	protocol := strings.ToLower(strings.TrimSpace(os.Getenv("OTEL_EXPORTER_OTLP_TRACES_PROTOCOL")))
	if protocol == "" {
		protocol = strings.ToLower(strings.TrimSpace(os.Getenv("OTEL_EXPORTER_OTLP_PROTOCOL")))
	}
	if protocol == "" {
		protocol = "http/protobuf"
	}

	switch protocol {
	case "http", "http/protobuf":
		exporter, err := otlptracehttp.New(context.Background())
		if err != nil {
			log.Printf(
				"[otel] failed to create OTLP HTTP exporter for %s: %v",
				streamingServiceName,
				err,
			)
			return nil
		}
		return exporter

	case "grpc":
		exporter, err := otlptracegrpc.New(context.Background())
		if err != nil {
			log.Printf(
				"[otel] failed to create OTLP gRPC exporter for %s: %v",
				streamingServiceName,
				err,
			)
			return nil
		}
		return exporter

	default:
		log.Printf(
			"[otel] unsupported OTLP traces protocol %q for %s; configure OTEL_EXPORTER_OTLP_TRACES_PROTOCOL=http/protobuf or grpc",
			protocol,
			streamingServiceName,
		)
		return nil
	}
}

func envFlag(name string, defaultValue bool) bool {
	raw := strings.TrimSpace(strings.ToLower(os.Getenv(name)))
	if raw == "" {
		return defaultValue
	}

	switch raw {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}

func randomHex(size int) string {
	buf := make([]byte, size)
	if _, err := rand.Read(buf); err != nil {
		return strings.Repeat("0", size*2)
	}
	return hex.EncodeToString(buf)
}

func attributesToKeyValues(attributes map[string]any) []attribute.KeyValue {
	items := make([]attribute.KeyValue, 0, len(attributes))

	for key, value := range attributes {
		switch typed := value.(type) {
		case bool:
			items = append(items, attribute.Bool(key, typed))
		case int:
			items = append(items, attribute.Int(key, typed))
		case int64:
			items = append(items, attribute.Int64(key, typed))
		default:
			text := asString(value)
			if text != "" {
				items = append(items, attribute.String(key, text))
			}
		}
	}

	return items
}

func asString(value any) string {
	switch typed := value.(type) {
	case string:
		return typed
	case int:
		return strconv.Itoa(typed)
	case int64:
		return strconv.FormatInt(typed, 10)
	case bool:
		if typed {
			return "true"
		}
		return "false"
	default:
		return ""
	}
}
